apis/Google.Cloud.Firestore.V1Beta1/Google.Cloud.Firestore.V1Beta1/FirestoreClientPartial.cs

```diff
@@ -0,0 +1,115 @@
+            ApplyResourcePrefixHeader(ref settings, request.Database);
+
+        partial void Modify_CommitRequest(ref CommitRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Database);
+
+        partial void Modify_CreateDocumentRequest(ref CreateDocumentRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Parent);
+
+        partial void Modify_DeleteDocumentRequest(ref DeleteDocumentRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Name);
+
+        partial void Modify_GetDocumentRequest(ref GetDocumentRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Name);
+
+        partial void Modify_ListCollectionIdsRequest(ref ListCollectionIdsRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Parent);
+
+        partial void Modify_ListDocumentsRequest(ref ListDocumentsRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Parent);
+
+        partial void Modify_RollbackRequest(ref RollbackRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Database);
+
+        partial void Modify_RunQueryRequest(ref RunQueryRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Parent);
+
+        partial void Modify_UpdateDocumentRequest(ref UpdateDocumentRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Document?.Name);
+
+        // TODO: Consider Write and Listen, where the settings and the requests are separate.
+
+        private static void ApplyResourcePrefixHeader(ref CallSettings settings, string resource)
+        {
+            // If we haven't been given a resource name, just let the request as it is.
+            if (string.IsNullOrEmpty(resource))
+            {
+                return;
+            }
+            string database = GetDatabaseResourceName(resource);
+            settings = settings.WithHeader(ResourcePrefixHeader, database);
+        }
+        
+        // Visible for testing
+
+        /// <summary>
+        /// Retrieves the database resource name from a full resource name.
+        /// Validation is performed as far as the database ID but no further; the database ID is deemed to end at the first slash.
+        /// </summary>
+        /// <param name="resource">The resource name, which must start with projects/{project_id}/databases/{database_id}</param>
+        /// <returns>The database resource name</returns>
+        internal static string GetDatabaseResourceName(string resource)
+        {
+            const string projectsPrefix = "projects/";
```

Comment:
This could probably all be done in a single regex.
Is this undesirable because it would be slower?

Reply:
That's a good question. Will benchmark, but yes, I suspect that would be slower.

Reply:
Now benchmarked: 50ns for manual code vs 300ns for regex.

Reply:
:)

Reply:
(Almost certainly insignificant in the grand scheme of things, but pretty easy to do, and it gets applied to all calls...)